### PR TITLE
WB-8581 log + link model finalize api design and implement end to end

### PR DESCRIPTION
**Pull request by vwrj**
https://wandb.atlassian.net/browse/WB-8581
https://wandb.atlassian.net/browse/WB-8597

Description
-----------
What does the PR do?

Testing
-------
How was this PR tested?

Checklist
-------
- Name PR "[WB-NNNN][WB-MMMM] Add support for..." similar to entries in CHANGELOG.md
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/beta/workflows.py
Comment (tssweeney) [inline wandb/beta/workflows.py:137]: I think in the future, we could extend this method to:
a) automatically do a `log_model` on the model if it doesn't belong to an artifact
b) allow the first parameter to be a path str, then automatically do the `use_model` under the hood

File: wandb/sdk/interface/artifacts.py
Comment (tssweeney) [inline wandb/sdk/interface/artifacts.py:665]: Another case of `registry` <> `portfolio` terminology to resolve

File: wandb/sdk/wandb_run.py
Comment (tssweeney) [inline wandb/sdk/wandb_run.py:2143]: Same as other one - probably want to use same logic as artifacts

File: wandb/sdk/wandb_run.py
Comment (tssweeney) [inline wandb/sdk/wandb_run.py:2140]: portfolio or registry ... the world may never know!

File: wandb/sdk/wandb_run.py
Comment (tssweeney) [inline wandb/sdk/wandb_run.py:2155]: Are you planning on getting this part in this PR?

**Review by tssweeney — CHANGES_REQUESTED** [review]
Overall looks really good - left comments for some various nits and improvements

File: wandb/apis/public.py
Comment (vwrj) [reply wandb/apis/public.py:3648]: fixed in next commit

File: wandb/apis/public.py
Comment (vwrj) [reply wandb/apis/public.py:3612]: changed in next commit

File: wandb/apis/public.py
Comment (vwrj) [reply wandb/apis/public.py:3643]: added in next commit

File: wandb/__init__.py
Comment (vwrj) [reply wandb/__init__.py:63]: Agreed, let's put everything into beta for now. Changed in next commit

File: wandb/util.py
Comment (vwrj) [reply wandb/util.py:1556]: `_parse_artifact_path` pulls entity and project from the settings if they don't exist but that wouldn't work for our use case, where the target portfolio might be under a different entity/project combination. I _could_make it work by modifying this function...but in any case, I don't like the idea of resolving entity and project inside the parse function. The parse function should only be focused on parsing the given path and extracting what it can, returning None or "" for the parts it can't find within the path. Which is why I created `_parse_entity_project_item`.

File: wandb/beta/__init__.py
Comment (vwrj) [reply wandb/beta/__init__.py:9]: Yep, I'm OK with having these methods exist at `wandb.beta.[namespace]`. I'll remove this file in the next commit. 

I don't have a strong opinion on the name -- I'll throw `model_workflows` into the mix :)

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:74]: deleted in next commit

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:124]: I'll change to `target_path` for now -- changed in next commit.

File: wandb/sdk/interface/artifacts.py
Comment (vwrj) [reply wandb/sdk/interface/artifacts.py:665]: changed to target_path in next commit

File: wandb/sdk/wandb_run.py
Comment (vwrj) [reply wandb/sdk/wandb_run.py:2140]: target for now, I suppose, to be generic :)

File: wandb/sdk/wandb_run.py
Comment (vwrj) [reply wandb/sdk/wandb_run.py:2155]: No, not in this PR.

File: wandb/sdk/wandb_run.py
Comment (vwrj) [reply wandb/sdk/wandb_run.py:2143]: Same response -- I'm not a fan of the artifact parse func because it pulls entity and project from the settings. That logic should be external to the function.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:20]: fixed in next commit

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:59]: added in next commit

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:205]: Yeah, this is interesting -- let's think about this. The value of `use_model` is two-fold:
1. For us, it binds the model artifact as an input to a run, which would ultimately allow us to do evaluation workflows in the UI. 
2. For the user, they get deserialization out of the box in an easy and framework-agnostic way, so they can do training and evaluation. 

Point 2 is always accomplished whether we start a run or not, so let's focus on 1. 

If we start a run for the user, 1 is accomplished _as long as the user does not do_ `wandb.init()` after the `use_model` call(s). If they do, however, the model artifacts won't be bound to the run we're interested in. 

If we throw an error (and make the user always init a run before `use_model`), I think this would ensure that 1 is accomplished correctly -- that is, the model artifacts are almost always bound to the same run that logs model metrics (i.e for the evaluation use-case).

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:210]: Agreed -- let's keep this as a stretch goal for now and otherwise rollover to the next phase.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:113]: added in next commit

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:202]: Nice; I'll add a try-except and checks for not None and that it's an instance of SavedModel. Added in next commit.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:173]: By gap, you mean literally the extra line right? Removing in next commit.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:247]: changed in next commit

**Review by tssweeney — APPROVED** [review]
Looks really good! Looking forward to getting this in PROD! One note: do not merge until the saved model pr is merged (and we have to change the PR base back)

File: functional_tests/artifacts/link-model.yea
Comment (kptkin) [inline functional_tests/artifacts/link-model.yea:8]: nit: are there other fields we can test here?

File: functional_tests/artifacts/log-model.py
Comment (kptkin) [inline functional_tests/artifacts/log-model.py:8]: nit: if it is the same model that being used for all test maybe it can be in a share file... obv pros and cons for both :)

File: functional_tests/artifacts/log-model.py
Comment (kptkin) [inline functional_tests/artifacts/log-model.py:37]: ah interesting... it creates a run implicitly?

File: wandb/apis/public.py
Comment (kptkin) [inline wandb/apis/public.py:3658]: general for this file: maybe add some interesting test cases for coverage?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:42]: can we add a test for this case?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:54]: this one should be easy to add a test

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:93]: do you want to do it implicitly or do you want to ask the user to start a new run?
for our other integrations we ask the user for this explicitly. I guess if we do it implicitly we should give the user some indication?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:96]: probably easy to hit this coverage

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:124]: in the args description you specify that the model obj can be either `PyTorch, Keras, Tensorflow, Scikit-learn` but here you also have `etc.` probably want to resolve the inconsistency?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:132]: so is `latest` a save alias?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:135]: oh so it could also belong to a run and a different project from the run's project?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:194]: wait do you need `:` between the path and the `latest`? 
```
use_model("my-simple-model")
```
will result in: `aliased_path = my-simple-modellatest`

from the description i'm not sure if this is a valid name...

Comment (kptkin) [reply wandb/beta/workflows.py:194]: probably good to add a simple test here

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:205]:
```suggestion
        if not isinstance(sm, (type(None), _SavedModel)):
```
but probably yours is more readable... so i would keep the current version

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:202]: do you want to be more specific with the exceptions?

File: wandb/beta/workflows.py
Comment (kptkin) [inline wandb/beta/workflows.py:216]: nit: do you want to have underscore as a public facing type?

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:19]: why do you need `six` we are only using py3+ ... `six` to allow support for py2... probably good idea to not introduce it in new code

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:415]: do you allow empty string for `server_id` in protobuf you could just not assign anything and check that this field exists, but it up to your implementation

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:2085]: nice :)

File: wandb/sdk/internal/internal_api.py
Comment (kptkin) [inline wandb/sdk/internal/internal_api.py:2101]: do you want to keep these logs?

File: wandb/sdk/wandb_run.py
Comment (kptkin) [inline wandb/sdk/wandb_run.py:2191]: probably easy to test this one... just get coverage points :)

File: wandb/sdk/interface/interface.py
Comment (kptkin) [inline wandb/sdk/interface/interface.py:420]: nit: 
```suggestion
        link_artifact.portfolio_aliases. extend(aliases)
```
see here:
https://developers.google.com/protocol-buffers/docs/reference/python-generated#repeated-fields

File: wandb/sdk/internal/sender.py
Comment (kptkin) [inline wandb/sdk/internal/sender.py:990]: is there a way to detect this sooner feels too late?

File: wandb/util.py
Comment (kptkin) [inline wandb/util.py:1581]: why not just reverse and extend with `[''] *(3-len(words))`?

File: wandb/util.py
Comment (kptkin) [inline wandb/util.py:1604]: do you need to check if aliases is of invalid type?

File: wandb/util.py
Comment (kptkin) [inline wandb/util.py:1574]: what would be a case when `len(words) == 0`? I think if it is an empty string it will be at least length 1?
maybe good idea to check that the path is of type `str`?

File: wandb/util.py
Comment (kptkin) [inline wandb/util.py:1581]:
```suggestion
    padded_words = [""] * (3 - len(words)) + words
    return tuple(reversed(padded_words))
```

File: functional_tests/artifacts/log-model.py
Comment (vwrj) [reply functional_tests/artifacts/log-model.py:8]: Nice -- I think 3 is the magic number. If I use it again in another test, I'l pull it into a shared file :)

File: functional_tests/artifacts/log-model.py
Comment (vwrj) [reply functional_tests/artifacts/log-model.py:37]: Yep, if `wandb.run` doesn't exist, we'll create a run for them. We definitely want to encourage users to create a run in the examples for this API, but include implicit run creation for user convenience.

File: functional_tests/artifacts/link-model.yea
Comment (vwrj) [reply functional_tests/artifacts/link-model.yea:8]: portfolio-links (line 9) should cover the test case; I can also add aliases but not sure how yea's assertion works with lists/arrays. Best thing for aliases would be set equality -- I don't think yea supports that yet, right?

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:132]: If you mean the user adding `latest` as an alias, it's safe because we make sure not to duplicate it.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:135]: ah, `project` is only used when creating a run implicitly on the user's behalf

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:194]: Ahh, great catch! Let me add this in the next commit. I think we want to make `:` required for `use_model` for now; let me raise an error otherwise. This may change in the future with user behavior; we may just allow the user to specify a name without an alias and we'll just pull either `best` if it exists or `latest`. Something like that.

File: wandb/sdk/interface/interface.py
Comment (vwrj) [reply wandb/sdk/interface/interface.py:19]: deleted in next commit

File: wandb/sdk/internal/internal_api.py
Comment (vwrj) [reply wandb/sdk/internal/internal_api.py:2101]: ah not really, deleted in next commit :)

File: wandb/sdk/interface/interface.py
Comment (vwrj) [reply wandb/sdk/interface/interface.py:420]: nice! I'll use this -- next commit.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:93]: I think we want it to work out of the box like a one-liner. So implicit for now, but you make a good point. This is under `wandb.beta.workflows` anyways so we can change it if users aren't comfortable with implicit creation.

File: wandb/beta/workflows.py
Comment (vwrj) [reply wandb/beta/workflows.py:124]: Those are going to be supported as of now, but we will most likely add more frameworks in the future, so keeping the method description flexible.

File: wandb/util.py
Comment (vwrj) [reply wandb/util.py:1605]: nice -- I'll add this check.

File: wandb/util.py
Comment (vwrj) [reply wandb/util.py:1575]: ah nice -- let me remove the `len(words) == 0` check.

File: wandb/util.py
Comment (vwrj) [reply wandb/util.py:1582]: because that would be a neat and easier solution :) -- changed in next commit

Comment (vwrj) [reply wandb/util.py:1582]: changed in next commit

**Pull request merged**
